src/PopIII/popiii.cpp

```diff
@@ -0,0 +1,479 @@
+
+		state_cc(i, j, k, HydroSystem<PopIII>::density_index) = state.rho;
+		state_cc(i, j, k, HydroSystem<PopIII>::x1Momentum_index) = state.rho * vx;
+		state_cc(i, j, k, HydroSystem<PopIII>::x2Momentum_index) = state.rho * vy;
+		state_cc(i, j, k, HydroSystem<PopIII>::x3Momentum_index) = state.rho * vz;
+		state_cc(i, j, k, HydroSystem<PopIII>::internalEnergy_index) = e;
+
+		Real const Egas = RadSystem<PopIII>::ComputeEgasFromEint(state.rho, state.rho * vx, state.rho * vy, state.rho * vz, e);
+		state_cc(i, j, k, HydroSystem<PopIII>::energy_index) = Egas;
+
+		for (int nn = 0; nn < NumSpec; ++nn) {
+			state_cc(i, j, k, HydroSystem<PopIII>::scalar0_index + nn) = mfracs[nn] * state.rho; // we use partial densities and not mass fractions
+		}
+	});
+}
+
+template <> void RadhydroSimulation<PopIII>::ErrorEst(int lev, amrex::TagBoxArray &tags, amrex::Real /*time*/, int /*ngrow*/)
+{
+	// refine on Jeans length
+	const int N_cells = 64; // inverse of the 'Jeans number' [Truelove et al. (1997)]
+	const amrex::Real G = Gconst_;
+	const amrex::Real dx = geom[lev].CellSizeArray()[0];
+
+	auto const &prob_lo = geom[lev].ProbLoArray();
+	auto const &prob_hi = geom[lev].ProbHiArray();
+
+	for (amrex::MFIter mfi(state_new_cc_[lev]); mfi.isValid(); ++mfi) {
+		const amrex::Box &box = mfi.validbox();
+		const auto state = state_new_cc_[lev].const_array(mfi);
+		const auto tag = tags.array(mfi);
+		const int nidx = HydroSystem<PopIII>::density_index;
+
+		amrex::ParallelFor(box, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+			amrex::Real const x = prob_lo[0] + (i + static_cast<amrex::Real>(0.5)) * dx;
+			amrex::Real const y = prob_lo[1] + (j + static_cast<amrex::Real>(0.5)) * dx;
+			amrex::Real const z = prob_lo[2] + (k + static_cast<amrex::Real>(0.5)) * dx;
+
+			Real const rho = state(i, j, k, nidx);
+			Real const pressure = HydroSystem<PopIII>::ComputePressure(state, i, j, k);
+			amrex::GpuArray<Real, Physics_Traits<PopIII>::numMassScalars> massScalars = RadSystem<PopIII>::ComputeMassScalars(state, i, j, k);
+
+			amrex::Real const cs = quokka::EOS<PopIII>::ComputeSoundSpeed(rho, pressure, massScalars);
+
+			const amrex::Real l_Jeans = cs * std::sqrt(M_PI / (G * rho));
+			if (l_Jeans < (N_cells * dx) && rho > 2e-20) {
+				tag(i, j, k) = amrex::TagBox::SET;
+			}
+		});
+	}
+}
+
+template <> void RadhydroSimulation<PopIII>::ComputeDerivedVar(int lev, std::string const &dname, amrex::MultiFab &mf, const int ncomp_cc_in) const
+{
+	// compute derived variables and save in 'mf'
+	if (dname == "temperature") {
+		const int ncomp = ncomp_cc_in;
+		auto const &state = state_new_cc_[lev].const_arrays();
+		auto output = mf.arrays();
+
+		amrex::ParallelFor(mf, [=] AMREX_GPU_DEVICE(int bx, int i, int j, int k) noexcept {
+			Real const rho = state[bx](i, j, k, HydroSystem<PopIII>::density_index);
+			amrex::Real const Eint = state[bx](i, j, k, HydroSystem<PopIII>::internalEnergy_index);
```

Comment:
You should compute this as Etot - KE instead.

Reply:
Why?

Reply:
You're right, it should be fine, since they are sync'd after each timestep.